### ISSUE TITLE
HttpClientFactory.createHttpClient() must produce AbstractHttpClient

### ISSUE THREAD

**Issue by johanhaleby** [jayway/rest-assured#497]
_From [sprag...@gmail.com](https://code.google.com/u/117167256599789367225/) on July 09, 2014 20:41:55_

\>> Note that currently you need to supply an instance of AbstractHttpClient. 

                    RestAssured.config = newConfig().httpClient(httpClientConfig().httpClientFactory(
                            new HttpClientConfig.HttpClientFactory() {
                                @Override
                                public HttpClient createHttpClient() {
                                    return new NotReallyAnHttpClient();
                                }
                            }));

Is this ever going to be fixed? I have an impl of HttpClient that is not an AbstractHttpClient. The cast to AbstractHttpClient fails.

_Original issue: http://code.google.com/p/rest-assured/issues/detail?id=342_

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on July 20, 2014 10:57:39_

I think It will require a massive rewrite. This is because http client has changed a lot in version 4.3 (I think). The HttpClientFactory API is prepared for the changed but I doubt that I'll find time to fix this in the near future. Feel free to help out if you like. Perhaps another couple of eyes would be good and hopefully find som trick to make it possible without a massive rewrite :)

**Status:** Accepted  
  **Labels:** -Type-Defect -Priority-Medium Type-Enhancement Priority-Low

**Comment by johanhaleby** [comment]
_From [lbeus...@googlemail.com](https://code.google.com/u/110491066608873617396/) on June 30, 2015 03:00:17_

I have the same problem and currently I can work around it. 

But the classes used internally by RestAssured are already deprecreated in the current httpclient version. So possible that they are removed in upcoming releases which causes RestAssured users to stick to the old version and stick to old versions of other libraries that depend on httpclient (which are a lot).

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on June 30, 2015 08:06:38_

What do you mean by workaround it? How do you do that?

**Comment by lbeuster** [comment]
Sorry for the late answer but I created the last post with the wrong mail address.

Workaround means only for my problem by using deprecated classes. I wanted to set the socket timeout and couldn't do that by using the regular http client builder - but that builder doesn't create an AbstractHttpClient. So I had to use the deprecated CoreConnectionPNames.

**Comment by johanhaleby** [comment]
Yeah I know this is bad.. Contributions are very welcome.

**Comment by eeichinger** [comment]
this issue is closely related to https://github.com/jayway/rest-assured/issues/548 - the HttpClient created by new HttpClientBuilder supports SNI, the deprecated DefaultHttpClient (which derives from AbstractHttpClient) does not

**Comment by johanhaleby** [comment]
@eeichinger The problem is that the newer versions of HttpClient is immutable. Once created REST Assured cannot add filters, encoders, authentication etc afaik. Thus RA depends on AbstractHttpClient which is mutable and allows you to do such things. Things get hard since changing the state of AbstractHttpClient is scattered more less throughout large parts of the code base. The only solution I can think of is to NOT allow people to set an instance of HttpClient but rather create a hook where people can access the HttpClientBuilder before it's created. RA probably needs to add its HTTP Client configuration before the builder is shipped off to the user hook. There are more problems as well  because if I remember it correctly there are multiple builders / HttpClient implementations and I don't think it'll be possible to support all of them. Also creation of HttpClient must be lazy and done just before the request is actually sent. Otherwise filters won't have the ability to change parameters, headers and other things before the request is sent. I'm sure that there are more issues that I haven't thought of as well.

I'm thinking that this may not be worth the effort. Perhaps we should focus on JDK 9 and work on REST Assured 3.0 which require Java 9 and uses the new built-in http client of Java 9 instead of Http Client? Not sure if it's a good idea and even feasible, but it could be an alternative.

**Comment by johanhaleby** [comment]
Just a note.

> There are more problems as well because if I remember it correctly there are multiple builders / HttpClient implementations and I don't think it'll be possible to support all of them.

Instead of maintaining a HttpClientBuilder internally in RA we could create an abstraction that is shipped to the "HttpClient instance creation hook". This way it's the responsibility to map from this abstraction into an HttpClient. This way it would be possible to use ANY http client (but if so it would also be the responsibility of the user to map from this abstraction to the HttpClient). There could be a middle ground here as well. We could for example supply two hooks, one where the user is responsible for the mapping entirely and another one where the user can hook into a default http client builder provided to the hook by RA (where RA does the mapping). The latter would probably solve most use cases and it would retain the ability to do more advanced configurations if really required. But all the hard problems still remains to be solved before we can start implementing this.

**Comment by sslavic** [comment]
If I'm not mistaken, rest-assured only has sync APIs (e.g. there's get but not getAsync or something), so then only (sync) HttpClientBuilder needs to be supported.

If considering to make an abstraction, why not make it so abstract that it's not tied to (API not exposing) any specific HTTP client implementation? Provide maybe one (sync Apache http client or OkHttp client) binding/implementation available out of the box.

**Comment by johanhaleby** [comment]
@sslavic That would of course be ideal. Problem is that I'm a single person with limited time :)

**Comment by sslavic** [comment]
From the two that I mentioned, which approach would you prefer, which makes sense from your persepective, first (sync apache http client only) or second (plugable high level abstraction)?

**Comment by johanhaleby** [comment]
Well, it would depend on how much work it would be to do the higher-level abstraction. What I was saying earlier that creating a REST Assured specific representation (interface) for all the HTTP Request and response state called `HttpClientBuilder` might be a good start. But I wonder if it would be much harder to just name it `RequestState` instead. So let me rephrase this a little bit to my latest thinking for an "ideal" solution. 

When using the DSL (given/when/then) I think that everything that gets specified ends up the `RequestState`. For example:

```java
given().param("x", "y").header("some", "value").when().post("/somewhere") .. 
```

would produce a `RequestState` with param "x"="y", header "some"="value" and resource "post" to "/somewhere".  So far there has been no interaction with any other library. So NOW comes the time to map the `RequestState` to a client specific library to actually send the requests. This library can be HttpClient and then REST Assured another lib could provide the mapping between the `RequestState` to an `HttpClient` instance that is wrapped in the REST Assured `Request` interface (psuedo code):

```java
(requestState) ->  new HttpClientRequest(new HttpClientBuilder().....build().setParam("x", "y").setHeader("some", "value"))
```

where `HttpClientRequest` is an `HttpClient` implementation of the `Request` interface. The `Request` interface could then have a method like this:

```java
sendRequest();
```

Which triggers the post to "/somewhere" using HttpClient in this case. This means that `rest-assured-core` could be extracted to not depend on any http library. Then we could create `rest-assured-apache-httpclient` that provides the mapping function.

BUT the cool thing with this is that users could plug other client libraries using the mapper function.

This, I (currently) think, would be the ideal solution. What do you think?

**Comment by kkocel** [comment]
@johanhaleby Hi, is this on the roadmap or anything has been done to create an abstraction layer over concrete HTTP clients?

**Comment by johanhaleby** [comment]
@kkocel This is something that would be great! Realistically, however, I won't have time to fix this in the foreseeable future. It will require a lot of rewriting.
